File: frontend/js/views/sde_valuation.js
// SDE business valuation — seller's discretionary earnings × multiple, via
// /calc/sde-valuation. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 }) + '%');

export async function renderSdeValuation(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.sde.h1.title">// SDE BUSINESS VALUATION</span></h1>
        <p class="muted small" data-i18n="view.sde.hint.intro">
            What an owner-operated business is worth. Buyers value Seller's Discretionary Earnings —
            net income with the owner's compensation and discretionary items added back — times an
            industry multiple (typically ~2–4×). The small-business analog of EV/EBITDA. Updates as
            you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.sde.h2.inputs">The business</h2>
            <form id="sde-form" class="inline-form">
                <label><span data-i18n="view.sde.label.ni">Net income ($)</span>
                    <input type="number" step="0.01" name="net_income_usd" value="100000" required></label>
                <label><span data-i18n="view.sde.label.owner">Owner compensation ($)</span>
                    <input type="number" step="0.01" min="0" name="owner_compensation_usd" value="80000" required></label>
                <label><span data-i18n="view.sde.label.da">Depreciation & amortization ($)</span>
                    <input type="number" step="0.01" min="0" name="depreciation_amortization_usd" value="15000"></label>
                <label><span data-i18n="view.sde.label.interest">Interest ($)</span>
                    <input type="number" step="0.01" min="0" name="interest_usd" value="5000"></label>
                <label><span data-i18n="view.sde.label.addbacks">Discretionary add-backs ($)</span>
                    <input type="number" step="0.01" min="0" name="discretionary_addbacks_usd" value="10000"></label>
                <label><span data-i18n="view.sde.label.multiple">SDE multiple</span>
                    <input type="number" step="0.1" min="0" name="sde_multiple" value="2.5" required></label>
                <label><span data-i18n="view.sde.label.revenue">Revenue ($)</span>
                    <input type="number" step="0.01" min="0" name="revenue_usd" value="600000"></label>
            </form>
        </div>
        <div id="sde-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#sde-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            net_income_usd: Number(fd.get('net_income_usd')) || 0,
            owner_compensation_usd: Number(fd.get('owner_compensation_usd')) || 0,
            depreciation_amortization_usd: Number(fd.get('depreciation_amortization_usd')) || 0,
            interest_usd: Number(fd.get('interest_usd')) || 0,
            discretionary_addbacks_usd: Number(fd.get('discretionary_addbacks_usd')) || 0,
            sde_multiple: Number(fd.get('sde_multiple')) || 0,
            revenue_usd: Number(fd.get('revenue_usd')) || 0,
        };
        try {
            const r = await api.calcSdeValuation(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.sde.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#sde-result');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.sde.h2.result">The valuation</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.sde.card.value">Business value</div>
                    <div class="value pos">${money(r.business_value_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.sde.card.sde">SDE</div>
                    <div class="value">${money(r.sde_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.sde.card.margin">SDE margin</div>
                    <div class="value">${pct(r.sde_margin_pct)}</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.sde.row.addbacks">Total add-backs</td><td>${money(r.total_addbacks_usd)}</td></tr>
                    <tr><td data-i18n="view.sde.row.sde">SDE</td><td>${money(r.sde_usd)}</td></tr>
                    <tr><td data-i18n="view.sde.row.margin">SDE margin</td><td>${pct(r.sde_margin_pct)}</td></tr>
                    <tr class="emph"><td data-i18n="view.sde.row.value">Business value</td><td>${money(r.business_value_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/sde_valuation.rs
//! Small-business valuation by Seller's Discretionary Earnings (SDE).
//!
//! For owner-operated businesses, buyers value the cash flow available to a
//! single working owner — net income with owner compensation and discretionary
//! items added back — times an industry multiple (typically ~2–4×).
//!
//! ```text
//! SDE = net income + owner comp + depreciation/amort + interest + add-backs
//! business value = SDE × multiple
//! ```
//!
//! This is the small-business analog of `ev-ebitda` for larger firms.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct SdeInput {
    pub net_income_usd: f64,
    /// Owner's salary/compensation (added back — a buyer sets their own).
    pub owner_compensation_usd: f64,
    #[serde(default)]
    pub depreciation_amortization_usd: f64,
    #[serde(default)]
    pub interest_usd: f64,
    /// Discretionary add-backs (owner perks, personal expenses, one-time items).
    #[serde(default)]
    pub discretionary_addbacks_usd: f64,
    /// Industry SDE multiple.
    pub sde_multiple: f64,
    /// Revenue, for the SDE margin. Optional.
    #[serde(default)]
    pub revenue_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct SdeResult {
    /// Seller's Discretionary Earnings.
    pub sde_usd: f64,
    /// SDE × multiple.
    pub business_value_usd: f64,
    /// SDE / revenue, percent; `None` if revenue ≤ 0.
    pub sde_margin_pct: Option<f64>,
    /// Total add-backs over net income (owner comp + D&A + interest + other).
    pub total_addbacks_usd: f64,
}

pub fn analyze(input: &SdeInput) -> SdeResult {
    let addbacks = input.owner_compensation_usd
        + input.depreciation_amortization_usd
        + input.interest_usd
        + input.discretionary_addbacks_usd;
    let sde = input.net_income_usd + addbacks;

    SdeResult {
        sde_usd: sde,
        business_value_usd: sde * input.sde_multiple,
        sde_margin_pct: if input.revenue_usd > 0.0 {
            Some(sde / input.revenue_usd * 100.0)
        } else {
            None
        },
        total_addbacks_usd: addbacks,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-4
    }

    fn base() -> SdeInput {
        SdeInput {
            net_income_usd: 100_000.0,
            owner_compensation_usd: 80_000.0,
            depreciation_amortization_usd: 15_000.0,
            interest_usd: 5_000.0,
            discretionary_addbacks_usd: 10_000.0,
            sde_multiple: 2.5,
            revenue_usd: 600_000.0,
        }
    }

    #[test]
    fn sde() {
        // 100k + 80k + 15k + 5k + 10k = 210,000.
        assert!(close(analyze(&base()).sde_usd, 210_000.0));
    }

    #[test]
    fn business_value() {
        // 210,000 × 2.5 = 525,000.
        assert!(close(analyze(&base()).business_value_usd, 525_000.0));
    }

    #[test]
    fn total_addbacks() {
        assert!(close(analyze(&base()).total_addbacks_usd, 110_000.0));
    }

    #[test]
    fn sde_margin() {
        // 210,000 / 600,000 = 35%.
        assert!(close(analyze(&base()).sde_margin_pct.unwrap(), 35.0));
    }

    #[test]
    fn higher_multiple_higher_value() {
        let low = analyze(&base());
        let high = analyze(&SdeInput {
            sde_multiple: 4.0,
            ..base()
        });
        assert!(high.business_value_usd > low.business_value_usd);
    }

    #[test]
    fn owner_comp_drives_addbacks() {
        let r = analyze(&SdeInput {
            owner_compensation_usd: 120_000.0,
            ..base()
        });
        assert!(close(r.sde_usd, 250_000.0));
    }

    #[test]
    fn zero_revenue_no_margin() {
        let r = analyze(&SdeInput {
            revenue_usd: 0.0,
            ..base()
        });
        assert!(r.sde_margin_pct.is_none());
    }

    #[test]
    fn no_addbacks_sde_equals_net_income() {
        let r = analyze(&SdeInput {
            owner_compensation_usd: 0.0,
            depreciation_amortization_usd: 0.0,
            interest_usd: 0.0,
            discretionary_addbacks_usd: 0.0,
            ..base()
        });
        assert!(close(r.sde_usd, 100_000.0));
    }
}
